datadog/fwprovider/resource_datadog_custom_framework.go

```diff
@@ -0,0 +1,358 @@
+package fwprovider
+
+import (
+	"context"
+	"strings"
+
+	"github.com/DataDog/datadog-api-client-go/v2/api/datadogV2"
+	"github.com/hashicorp/terraform-plugin-framework-validators/setvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/utils"
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/validators"
+)
+
+var _ resource.Resource = &customFrameworkResource{}
+
+type customFrameworkResource struct {
+	Api  *datadogV2.SecurityMonitoringApi
+	Auth context.Context
+}
+
+type customFrameworkModel struct {
+	ID           types.String `tfsdk:"id"`
+	Description  types.String `tfsdk:"description"`
+	Version      types.String `tfsdk:"version"`
+	Handle       types.String `tfsdk:"handle"`
+	Name         types.String `tfsdk:"name"`
+	IconURL      types.String `tfsdk:"icon_url"`
+	Requirements types.Set    `tfsdk:"requirements"` // have to define requirements as a set to be unordered
+}
+
+func NewCustomFrameworkResource() resource.Resource {
+	return &customFrameworkResource{}
+}
+
+func (r *customFrameworkResource) Metadata(_ context.Context, _ resource.MetadataRequest, response *resource.MetadataResponse) {
+	response.TypeName = "custom_framework"
+}
+
+func (r *customFrameworkResource) Schema(_ context.Context, _ resource.SchemaRequest, response *resource.SchemaResponse) {
+	response.Schema = schema.Schema{
+		Description: "Manages custom framework in Datadog.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Description: "The ID of the custom framework resource.",
+				Computed:    true,
+			},
+			"version": schema.StringAttribute{
+				Description: "The framework version.",
+				Required:    true,
+			},
+			"handle": schema.StringAttribute{
+				Description: "The framework handle.",
+				Required:    true,
+			},
+			"name": schema.StringAttribute{
+				Description: "The framework name.",
+				Required:    true,
+			},
+			"icon_url": schema.StringAttribute{
+				Description: "The URL of the icon representing the framework.",
+				Optional:    true,
+			},
+			"description": schema.StringAttribute{
+				Description: "The description of the framework.",
+				Optional:    true,
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"requirements": schema.SetNestedBlock{
```

Comment:
Can't set a nested block Requirements required as true and can't set nested block Controls required as true. When generating the docs Controls is set as optional due to this.